### ISSUE TITLE
Possible to declare extern method/function with return type 'int'?

### ISSUE THREAD

**Issue by jfingerh** [p4lang/p4c#2475]
I have a use case where it would be nice to have an extern method or function in an architecture that returns a compile-time known value of type 'int', and where its return value could be used in places where only a compile-time known value is allowed, e.g. as a parameter to another extern object constructor call.

I tried compiling the attached program using the latest version of p4c as of 2020-Jul-22, but am getting error messages like these:
```
$ p4c --target bmv2 --arch v1model tmp1.p4
tmp1.p4(55): [--Werror=type-error] error: my_fn1: Cannot unify type int with return type
    bit<32> tmp1a = (bit<32>) my_fn1(5);
                              ^^^^^^^^^
tmp1.p4(56): [--Werror=type-error] error: my_fn1: Cannot unify type int with return type
    bit<32> tmp1b = (bit<32>) my_fn1((bit<32>) 5);
                              ^^^^^^^^^^^^^^^^^^^
tmp1.p4(57): [--Werror=type-error] error: my_fn1: Cannot unify type int with return type
    custom_extern1(my_fn1(5)) e4;
                   ^^^^^^^^^
```
For example, the built-in method `minSizeInBytes()` is not explicitly declared in core.p4, but if one were to explicitly declare it, its return type seems to me like it could be defined as type `int`, and is a compile-time known value.
[tmp1.p4.txt](https://github.com/p4lang/p4c/files/4962241/tmp1.p4.txt)

**Comment by mbudiu-vmw** [comment]
The front end does not know how to evaluate such a function, so it cannot replace it with a constant.

**Comment by jafingerhut** [comment]
Understood, and agree that the front end cannot replace it with a constant.

Would you consider p4c to be designed in such a way that it would need significant redesign in order to support functions that return compile-time known values, whose return values are not known until the back end?

**Comment by mbudiu-vmw** [comment]
It should not be a major redesign, we just have to agree on an API by which the back-end can inform the front-end about needed features of the architecture. Some generalization of the policies we use in the mid-end passes, but with a single uniform API for all back-ends.

**Comment by jafingerhut** [comment]
I am not against a single uniform API as you describe, but out of curiosity, is any new API needed in order for the front end to know that something with a return type of `int` must return a compile-time known value?

**Comment by mbudiu-vmw** [comment]
No, but it needs to actually compute the value. The front-end assumes that compile-time known values are known and participate, for example, in constant folding.

**Comment by jfingerh** [comment]
If there were a separate additional constant folding pass later in the mid/back end, would that be sufficient?  Or does the front end make additional assumptions?

**Comment by mbudiu-vmw** [comment]
It's hard to tell without trying.
Certainly, currently the type checker is upset if it cannot infer the width of a runtime expression at compile-time, i.e., the return type of the function in your case. This we can probably fix. We could require that all objects that are assigned an int value must be `const`, maybe that's enough.

**Comment by jfingerh** [comment]
You say: " currently the type checker is upset if it cannot infer the width of a runtime expression at compile-time, i.e., the return type of the function in your case."

In my test program, which runtime expression is p4c trying to infer the width of?  (perhaps the only example of such things are ones that I commented out because they produce errors from p4c)

**Comment by mbudiu-vmw** [comment]
It wants to know how wide the return value of the function is. Currently it assumes that functions are never executed at compile-time.
